### ISSUE TITLE
Parsing performance

### ISSUE THREAD

**Issue by DocKDE** [douweschulte/pdbtbx#81]
I've stunbled upon PDB parsing libraries in other languages (specifically Haskell, [hPDB](https://github.com/BioHaskell/hPDB)) and noticed that they advertise speed of parsing to be very high. 
This PDB file (https://www.rcsb.org/structure/1HTQ) is apparently parsed in under 10 seconds (I didn't verify this though). I used `pdbtbx` to read the very same file and it took 20 minutes with a release build.
This leaves me wondering whether there is room for improvement when it comes to the parsing of PDB files in `pdbtbx`.
I realize this is not a very concrete issue but I wanted to ask for your thoughts on this.

**Comment by douweschulte** [comment]
Interesting point, the (very limited) comparison I did between cctbx (C based used with python api) and pdbtbx came to similar speeds, but with much smaller PDBs. There is always room for improvement, but I do not right now know how to get there. Although I would take a look into the places where unnecessary new memory is used and in the way of adding new atoms to the list. I assume the really fast libraries have fancier ways of finding the right conformer to add the atom to. Also I see now that the `1HTQ` has 10 models of nearly 100 000 atoms, I think that in that case the verification also takes a long while.

So yes nice point, I will do some measuring and see if there is something obvious I can improve. It would be a shame to use an extremely fast language and have a slower library than a (in normal cases) slower language ;-).

**Comment by douweschulte** [comment]
Sidenote I think this is a great resource if you ever want to do some performance work in Rust: https://nnethercote.github.io/perf-book/introduction.html or another book https://www.packtpub.com/product/rust-high-performance/9781788399487.

**Comment by DocKDE** [comment]
I agree, it would be a shame that's why I opened this issue. I did some few optimizations in my own crate and I'm mostly happy with its performance. The step taking the longest by far is the actual parsing :)

**Comment by DocKDE** [comment]
I ran a quick flamegraph analysis on a simple executable which does nothing but opening a PDB file. Apparently, the vast majority of time is spent within the `chain::add_atom` function. Why exactly that is, I don't know (yet)

**Comment by douweschulte** [comment]
Nice! I would assume (definitively needs some measurements) that it takes a long time to find the correct residue + conformer to add to. If this is the case I think starting from the end of the list with the search for the correct res + conf is faster.

**Comment by DocKDE** [comment]
I investigated some more. 
The binary I was looking at contains just this code:
```
fn main() {
    let _a = pdbtbx::open_pdb("large.pdb", pdbtbx::StrictnessLevel::Strict);
}
```
The PDB file is the same as in the `pdbtbx` example files which I used to test for files with more than 100k atoms. I've never profiled anything before so I don't really know much about what I'm doing but I used `perf` to take recordings from running this binary and determined (as with the flamegraph before) that most of the time is spent within the `chain::add_atom` method. 
Looking at the generated assembler code, I got this:
```
  0,19 │ c0:   add    rbp,0x38                                                                                   
  0,05 │       add    r12,0xffffffffffffffc8                                                                   
       │     ↓ je     114                                                                                       
  1,00 │ ca:   cmp    QWORD PTR [rbp+0x0],rbx                                                                      
 98,60 │     ↑ jne    c0                                                                                         
  0,01 │       mov    rdi,QWORD PTR [rbp+0x8]                                                                     
       │       test   r13,r13                                                                                    
       │       setne  al                                                                                          
```
I have to confess that I don't know anything about assembly but the hottest instruction is `jne` which is a "jump if not equal" one. If you happen to know more, I'd love to hear your thoughts.
I reasoned that, according to this finding, the bottleneck was the evaluation of one of the `if` statements in the method. I commented out some stuff, recompiled, ran the binary again and found that the code evaluation that takes most of the time (by a large margin) is the two lines within the second `if` statement:
```
if !found {
    self.residues.push(new_residue);
    current_residue = self.residues.last_mut().unwrap();
}
```
Also, the bottleneck is not the evaluation of `found` but rather the instructions within the body. Since these two depend on one another and/or other things, I couldn't narrow it down further for the time being.

This shows, however, that the `for` loop determining whether the current residue is already present is, in fact, not the problem, unless I'm much mistaken here.

**Comment by DocKDE** [comment]
Okay, I dug a little more into profiling the performance and I think I've come to the bottom of the issue now. What I wrote earlier seems to have been an incorrect assessment of what the issue was.
I performed a couple runs with `perf` on a binary compiled from this:
```
fn main() {
    let _a = pdbtbx::open_pdb("large.pdb", pdbtbx::StrictnessLevel::Strict);
}
```
Then I recorded a profile with `perf record --call-graph=dwarf ./target/release/pdbtest` and looked at it with `perf report`. Looking at the call stack, the conclusion that the `chain::add_atom` method is the culprit, however it appears that it, in turn, calls this: `core::tuple::<impl core::cmp::PartialEq for (A,B)>::eq (inlined)` which makes up some 88% of the time spent during run time.
So it seems clear that the bottleneck is the comparison `if residue.id() == residue_id` within the `add_atom` method. 
Since the comparison has already been inlined by the compiler, I'm not sure how to improve the situation just yet but I wanted to share the information.
If you have any criticisms or suggestions, I'm all ears!

**Comment by DocKDE** [comment]
I created a PR that adds some clippy lints and the reversal of iteration that you mentioned. Turns out your assessment of what the bottleneck is, was exactly right. :)
#82

**Comment by douweschulte** [comment]
The comparison is done when searching for the right residue when inserting a new one. By going from the back in most cases ((n-1)/n where n is the average number of atoms in a residue) the residue is found immediately, but if the residue has not been added yet it still has to go through the whole list. If we assume that no residue is ever defined with another residue in between, the best way forward would be the same trick as chains in the parser. Keep the new residue local in the parser and add all atoms until a new residue is found. At this point the residue has to be added to the current chain and a new one has to be made.
So that would be a huge improvement as well I assume, but I am not entirely sure this assumption holds :/.

If it does not hold some kind of set made for constant lookup time with just the residue ids could be beneficial by removing the need for the full comparison loop.

**Comment by DocKDE** [comment]
I was thinking along the same lines just now (I think). If I understand correctly, each time an atom is to be added (for `large.pdb` that would be more than 100k times) the entire current chain is traversed in search for a residue to add the atom to.
So each time an atom of a new residue is added, the whole chain is searched in vain which would amount to a lot of pointless searching. I hope I got that right, I tried to use my own words for it :D

My suggestion would be to introduce either a set holding the already present residue IDs or do some sort of memoization via a HashMap to store residues that have already been added. 
Either way, this should reduce lookup times for residue IDs rather drastically. Probably the set approach would be easier to do, I have already used sets with success in my own `pdbman` project to reduce the traversal of `Vec`s.

**Comment by douweschulte** [comment]
Yes that is exactly what I was thinking about. I think that changing the internal storage from `Vec` to some other structure is justified for this cause. That would probably also improve performance in some other cases.

**Comment by DocKDE** [comment]
I would suggest using an `IndexSet` from the `indexmap` crate (https://docs.rs/indexmap/latest/indexmap/set/struct.IndexSet.html). It's an ordered set which has implementations for many standard methods for `Vec`s but features constant-time lookups. I think it should be a good fit because Residue and Atom IDs are supposed to be unique anyway.
In principle, other internals that are now represented with `Vec`s could be switched to such sets as well, I think.

**Comment by DocKDE** [comment]
A possible pitfall (which I've previously come across) is that storing `Residue`s, `Atom`s and the like in `HashSet`s or `IndexSet`s won't work out of the box because they all incorporate `f64` fields which don't implement `Hash`. A possible solution here would be using `OrderedFloat`s from here (https://docs.rs/ordered-float/latest/ordered_float/struct.OrderedFloat.html).
I suppose all of this would require some major refactoring but I think it should lead to significant speedups during parsing.

**Comment by douweschulte** [comment]
Implementing `Hash` is doable actually. The equality is defined by the serial_numbers/names for each level so for floats are involved in that. Creating a commit implementing `Hash` for everything.

**Comment by DocKDE** [comment]
I had a look at this myself and came to the conclusion that implementing `Hash` on `Atom` (which is the only one that can't just be derived) only works if you exclude all of the fields that are floats.
For example one could hash a tuple of name, serial number and so on but not including the position, occupancy and b_factor etc.
Technically this could then lead to logic errors because the criteria for equality and hash equality would not be the same. In practice that's not very likely, I think.

**Comment by DocKDE** [comment]
Additional note: I just realized that using sets as drop-in replacement for the `Vec`s that are currently used needs some more thought because creating a mutable reference to a set item is not possible. It should be possible to create a workaround but that may come at additional cost, especially if the order of the set is to be preserved.

**Comment by douweschulte** [comment]
Okay I tried some things out myself, including the very nice `IndexMap` you pointed out earlier. And I think it will prove to be very hard to keep a set of some sort in permanent use, as it will be nearly impossible to keep all data in sync especially if you want to keep the opportunity for people to change the id of residues. So I think the approach you took in #86 has the most potential. Here you create a set for single usage, only while parsing, removing many possibilities for out of sync data. The implementation then could be in the form as you did, an extra function which takes the `HashSet` of already added `Residues`. Or there could be an intermediate `HashMap` while parsing which is turned into a `Chain` as soon as a new chain is detected.

**Comment by DocKDE** [comment]
Sure, looking forward to it!
